### DISCUSSION TITLE
Need better permissions for GitHub Packages from GitHub Actions

### DISCUSSION THREAD

**Discussion by chriskuech** [community/community#9840] [Actions]
The main benefit of private artifact feeds like GitHub Packages is distributing packages internally in an organization.  Currently this is very difficult in GitHub Actions because GITHUB_TOKEN is repository scoped.  While GITHUB_TOKEN should not ever be able to write another repository, it should be allowed to read any repository's packages within the same Team/Organization.

Similarly, an Action in one repository should be allowed to trigger off a Package publication in another repository, such as to allow automation related to package updates.

**Comment by pulkitaggarwl** [comment]
Thanks for reaching out.  We have added support for the org-level container package which means that you can use GITHUB_TOKEN to pull and push images that have visibility outside the repo.  One can set the [visibility ](https://docs.github.com/en/packages/learn-github-packages/configuring-a-packages-access-control-and-visibility#container-creation-visibility-for-organization-members) of the package to internal, public, or private. If visibility is set to internal another repo will be able to pull the package using GITHUB_TOKEN.  We are working on adding this functionality to other packages types. 

> Similarly, an Action in one repository should be allowed to trigger off a Package publication in another repository, such as to allow automation related to package updates. 

When you say "Action" I am assuming GH Actions as part of GH workflow. If so, you can create a [repository dispatch](https://docs.github.com/en/actions/learn-github-actions/events-that-trigger-workflows#repository_dispatch) event that triggers the alternate workflow where you are publishing the package. Have you tried that?

**Reply by chriskuech** [reply]
Thanks @pulkitaggarwl , but I don't think repository dispatch would work for my scenario due to the below issue:

> When you use the repository's GITHUB_TOKEN to perform tasks, events triggered by the GITHUB_TOKEN will not create a new workflow run. This prevents you from accidentally creating recursive workflow runs. For example, if a workflow run pushes code using the repository's GITHUB_TOKEN, a new workflow will not run even when the repository contains a workflow configured to run when push events occur. For more information, see "Authenticating with the GITHUB_TOKEN."

[source](https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow)

**Reply by YouriT** [reply]
> We are working on adding this functionality to other packages types.

Any timeline in that regard? Any order of rollout?

What's the workaround to fetch a package (npm, nuget, etc) from another repo? Creating a PAT and inserting it as a secret in the pipeline?

**Reply by chriskuech** [reply]
~Another issue I'm running into is just reading files with the API across repos~

**Reply by chriskuech** [reply]
It's also unclear what happens if I use a PAT token in an organization workflow.
* Does the PAT have access to multiple organizations?
* Does the PAT have access to my private repositories?

**Reply by pulkitaggarwl** [reply]
> Thanks @pulkitaggarwl , but I don't think repository dispatch would work for my scenario due to the below issue:

> When you use the repository's GITHUB_TOKEN to perform tasks, events triggered by the GITHUB_TOKEN will not create a new workflow run. This prevents you from accidentally creating recursive workflow runs. For example, if a workflow run pushes code using the repository's GITHUB_TOKEN, a new workflow will not run even when the repository contains a workflow configured to run when push events occur. For more information, see "Authenticating with the GITHUB_TOKEN."
> What's the workaround to fetch a package (npm, nuget, etc) from another repo? Creating a PAT and inserting it as a secret in the pipeline?

The @chriskuech  for sharing the details, you can use the PAT token in the interim. I see that there are some questions around PAT token, have answered them below. In case you need more details OR PAT token route doesn't work for you, let us know. 

> Another issue I'm running into is just reading files with the API across repos

Please elaborate the challenges you are facing while reading files via APIs, it will help understand the situation better.


> Does the PAT have access to multiple organizations?
> Does the PAT have access to my private repositories? 

Yes, if the user that is generating the PAT token belongs to have access to multiple orgs and your private repos or any other repos for that matter.  The PAT token mirrors the access permissions of the user that generated the token with a coarse permission model around the resources you enable the PAT for (e.g. repository read, read:packages, etc…). If you enable repo (Full control of private repositories) for the PAT, then any private repository the user has access to will be provided via the PAT. This extends to organizations as well.
Having said that, If the organizations are protected via SSO, then the[ user must authroize this PAT to provide access to those organizations](https://docs.github.com/en/enterprise-cloud@latest/authentication/authenticating-with-saml-single-sign-on/authorizing-a-personal-access-token-for-use-with-saml-single-sign-on). If SSO is not in place, then the PAT token will gain access implicitly.

**Reply by chriskuech** [reply]
> Another issue I'm running into is just reading files with the API across repos

Thanks for the followup.  This was some user error on my end, trying to use `GITHUB_TOKEN` env var with `gh api` rather than using `gh auth` first.  Please disregard.

**Reply by chriskuech** [reply]
Thanks @pulkitaggarwl .  I think I am unblocked by using PATs but it's pretty annoying that there is no isolation across accounts.  My primary feedback request following this thread is that PATs can be scoped to a single account (organization or personal user) or even a Team, to really unlock the potential of GitHub Actions for complex automation.

**Reply by jsoref** [reply]
The standard advice is to create a bot user per org (yes it costs a seat) and issue PATs from it. That way the PATs don't have access to other orgs to which a given user may have access. It also means that the account will survive any particular user leaving an org.

**Reply by dominik-ba** [reply]
> The standard advice is to create a bot user per org (yes it costs a seat) and issue PATs from it. That way the PATs don't have access to other orgs to which a given user may have access. It also means that the account will survive any particular user leaving an org.

is there such a thing as a bot user in GitHub and I don't find it or you mean creating a real new user which credentials you share? 🙀

**Reply by jsoref** [reply]
I mean a normal user that just happens to have `-bot` in its name.

The other approach is to set up an OAuth App or a GitHub App. But I've never made my own, so I have less experience with it and can't really say how it might work.
